spices/0002-uri-rewriting-for-mirroring.adoc

```diff
@@ -0,0 +1,319 @@
+When dependencies can not be determined ahead of time, pre-populating the cache is not an option.
+Mirroring packages from internet repositories in a local store is insufficient, because the mirrored packages may have hard-coded dependencies.
+Importing mirrored packages, using URIs pointing to the mirror, will eventually evaluate `import`s of those hard-coded dependencies, referring to external sites.
+
+Another reason to want to consume (transitive) dependencies from a mirror is the requirement of altered versions of packages.
+For example, redirecting a package developed for `github.com` to point to a GitHub Enterprise instance.
+Note that, because upstream dependencies specify the SHA256 hash, altering packages on a mirror requires _all upstream dependants to be updated_.
+This SPICE does _not_ provide further facilities for altering packages when mirroring.
+
+== Proposed Solution
+
+Similar to `git-config`'s `insteadOf`, the proposed solution is to rewrite URIs when used in `import`, `import*`, `amends`, `extends`, `read`, and `read*` clauses/expressions.
+From `git-config`'s `man` page, on `url.<base>.insteadOf`:
+
+> Any URL that starts with this value will be rewritten to start, instead, with `<base>`.
+> In cases where some site serves a large number of repositories, and serves them with multiple access methods, and some users need to use different access methods, this feature allows people to specify any of the equivalent URLs and have Git automatically rewrite the URL to the best alternative for the particular user, even for a never-before-seen repository on the site.
+> When more than one insteadOf strings match a given URL, the longest match is used.
+>
+> Note that any protocol restrictions will be applied to the rewritten URL.
+> If the rewrite changes the URL to use a custom protocol or remote helper, you may need to adjust the `protocol.*.allow` config to permit the request.
+
+The solution proposed in this SPICE mirrors this design, but extends it to allow regular expressions for the match.
+Only after rewriting are URIs tested for compliance with Allowed Modules and Allowed Resources settings.
+
+Rewrites can be provided:
+
+1. Via the API;
+2. Via `pkl:Project`; and
+3. Via arguments on the CLI.
+
+== Detailed design
+
+=== API
+
+The API changes described here refer to the Java API, but this SPICE implies analogous changes to the Swift and Go APIs.
+
+The `EvaluatorBuilder` gets the following new methods:
+
+[source,java]
+----
+public final class EvaluatorBuilder {
+    // ...
+
+    /**
+     * Adds the given URI rewrite rule.
+     *
+     * @param replace   The literal string to match.
+     * @param with      The literal string to substitute for the first match of [replace].
+     */
+    public EvaluatorBuilder addRewriteRule(String replace, String with) {
```

Comment:
I think the method name should be more specific. Suggestions: `addModuleUriRewriteRule`, `addImportUriRewriteRule’, ‘rewriteImportUri`. When I see `String replace`, I’m thinking that’s the replacement.